modules/references/tests/server/utils.js

```diff
@@ -0,0 +1,165 @@
+ * @param {boolean} [defs.publ] - is the user public? (defaults to a random boolean)
+ * @param {string} [defs.password] - password (defaults to a random password)
+ * @returns {object[]} array of user data
+ */
+function generateUsers(count, defs) {
+  defs = _.defaultsDeep(defs, {
+    username: 'username',
+    firstName: 'GivenName',
+    lastName: 'FamilyName',
+    email: 'user@example.com'
+  });
+
+  return _.range(count).map(function (i) {
+    var username = defs.username + i;
+    var firstName = defs.firstName + i;
+    var lastName = defs.lastName + i;
+    var email = i + defs.email;
+    var publ = defs.hasOwnProperty('public') ? defs.public : !getRandInt(2); // public is reserved word
+    var password = defs.password || crypto.randomBytes(24).toString('base64');
+
+    return {
+      public: publ,
+      firstName: firstName,
+      lastName: lastName,
+      email: email,
+      username: username,
+      displayUsername: username,
+      password: password
+    };
+  });
+}
+
+/**
+ * Generate reference objects.
+ * @param {object[]} users - array of mongodb User
+ * @param {[number, number, object][]} referenceData - array of data for each reference
+ * @param {number} referenceData[][0] - index of userFrom in users
+ * @param {number} referenceData[][1] - index of userTo in users
+ * @param {object} referenceData[][2] - object of property: value to override default reference properties
+ */
+function generateReferences(users, referenceData) {
+  return referenceData.map(function (data) {
+    var defaultReference = {
+      userFrom: users[data[0]]._id,
+      userTo: users[data[1]]._id,
+      public: true,
+      met: !getRandInt(2),
+      hostedMe: !getRandInt(2),
+      hostedThem: !getRandInt(2),
+      recommend: ['yes', 'no', 'unknown'][getRandInt(3)]
+    };
+
+    return _.defaultsDeep({}, data[2], defaultReference);
+  });
+}
+
+/**
+ * @callback {saveDocumentsCallback}
+ * @param {error|null} error
+ * @param {object[]} documents - array of saved mongo documents
+ */
+
+/**
+ * Save documents to mongodb
+ * @param {string} collection - name of mongoose model (mongodb collection)
+ * @param {object[]} _documents - array of document data
+ * @param {saveDocumentsCallback} done
+ */
+function saveDocumentsToCollection(collection, _docs, done) {
+  var docs = _docs.map(function (_doc) {
+    var Model = mongoose.model(collection);
+    return new Model(_doc);
+  });
+
+  async.eachSeries(docs, function (doc, cb) {
+    doc.save(cb);
+  }, function (err) {
+    return done(err, docs);
+  });
+}
+
+var saveUsers = _.partial(saveDocumentsToCollection, 'User');
+var saveReferences = _.partial(saveDocumentsToCollection, 'Reference');
+
+/**
+ * Clear specified database collections
+ * @param {string[]} - array of collection names to have all documents removed
+ * @param {function} - callback
+ */
+function clearDatabase(collections, done) {
```

Comment:
I've been thinking we could just nuke all the collections from test DB after each test. It would simplify things a bit. What do you think?

Reply:
So that would mean getting a list of all collections and then delete documents in all of them?

i.e. getting a list from `mongoose.connection.db.collections()`? Or do you imagine another way?

Reply:
> I think we could just have a fixed list like we basically have now, but in one place all at once? Would that be confusing? My rationale is that it would make each set of tests little bit more simple and less to worry about when writing or updating tests.

Sounds good.

_meta:_ I think we're experiencing a github user experience bug here. A review comment sent by email doesn't end up in the right place.

Reply:
Replaced the collection parameter with a fixed list of collections. df7be55

The list is incomplete and will need to be extended as needs arise.